added values for HSMETRICS_QC_CHECK

keyvanelhami opened the pull request:
### This PR:

This PR addresses issue #394 .

Added: New dictionary is added where the required QC parameters for the Balsamic analysis to pass, are added with different values for different bait sets. 


### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

keyvanelhami commented:
@hassanfa Please review the newly added "HSMETRICS_QC_CHECK" in the j.son file. Does it looks like a good procedure?

hassanfa commented:
Awesome! What do you think about having versions in panel bed names? I mean it can be useful to have it so we can track them back. But I don't think it matters which panel version it is, what do you think?

keyvanelhami commented:
@hassanfa I think it's important to include the bed versions, and I have already added it to the bed name, e.g. `gmck_solid_4.1`. Or do you mean something else with bed versions?

hassanfa commented:
Been thinking about this for a while, and if we need to have control over the version for panel, we could use a structure such as the following:
```
QC_METRICS = {
	'gi_cfdna_3.1': {
		'MEAN_TARGET_COVERAGE': 500,
		'FOLD_80_BASE_PENALTY': 1.5,
		'PCT_OFF_BAIT': 0.35
	},
	'wes': {
		'MEAN_TARGET_COVERAGE': 500,
		'FOLD_80_BASE_PENALTY': 1.5,
		'PCT_OFF_BAIT': 0.35
	},
	'wgs': {
		'FOLD_80_BASE_PENALTY': 1.5,
		'PCT_OFF_BAIT': 0.35
	}
}
```

We could mix and match QC_METRICS based on panel we have. So key names will be panel names, or application type. It could be WGS, WES, RNA, or even specific panel name such as `gi_cfdna_3.1`.

ashwini06 commented on BALSAMIC/utils/constants.py:
```diff
@@ -42,7 +42,48 @@
 
+# Minimum required QC-values from HS metrics to be able to pass analysis
+HSMETRICS_QC_CHECK = {
+    'gi_cfdna_3.1': {
```
just for consistency 'gi_cfdna_3.1' -> "gi_cfdna_3.1"

ashwini06 commented on BALSAMIC/utils/constants.py:
```diff
@@ -42,7 +42,48 @@
+        "PCT_OFF_BAIT": 0.25,
+    },
+    "wgs": {
+        "MEAN_TARGET_COVERAGE": 30,
```
Is other params missing here for wgs?

hassanfa replied:
Agreed, we need to expand WGS features.

ashwini06 reviewed (approved):
few comments

hassanfa commented:
I changed the base branch to develop from master.

keyvanelhami commented:
@ashwini06 @hassanfa I pushed a python script `qc_check.py` that provides a table with qc check for samples from the HS metrics. I know there is room for a lot of improvements. Please have a look at the script and provide som feedback.

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+import click
+
+
+qc_list = ['MEAN_TARGET_COVERAGE', 'FOLD_80_BASE_PENALTY', 'PCT_OFF_BAIT']
```
1. Move this to `BALSAMIC/utils/constants.py` and then import it here: `from BALSAMIC.utils.constants import QC_LIST`
1. Lets have a descriptive variable name here. And being a constant/global, lets have it upper case `qc_list` --> `BASE_QUALITY_METRICS`

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+
+def read_hs_metrics(hs_metrics_file):
+
```
```suggestion
```

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+qc_list = ['MEAN_TARGET_COVERAGE', 'FOLD_80_BASE_PENALTY', 'PCT_OFF_BAIT']
+
+
+def read_hs_metrics(hs_metrics_file):
```
Add type hinting:
```suggestion
def read_hs_metrics(hs_metrics_file: str):
```

keyvanelhami replied:
How his type hinting used for DataFrames?

hassanfa replied:
example: 
```
import pandas

def some_function(my_df: pandas.DataFrame) -> pandas.DataFrame:
     ...

```

hassanfa replied:
`-> pandas.DataFrame` is to typehint return.

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+def read_hs_metrics(hs_metrics_file):
+
+    '''
+    input: string
+    output: DataFrame
+
+    Reads the HS metrics (json-format) and returns it as a DataFrame
+    '''
```
Please use the following format for all other functions below.

```suggestion
    ''' Reads the HS metrics (json-format) and returns it as a DataFrame
    
    Args:
        hs_metrics_file: A string path to the file

    Returns:
        metrics_df: DataFrame
    '''
```

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+
+def read_qc_table(qc_table_file):
+
```
```suggestion
```

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+def read_qc_table(qc_table_file):
+
+    '''
+    input: string
+    output: DataFrame
+
+    Reads the QC-table (json-format) and returns it as a DataFrame
+    '''
```
Similar to previous functions, please update:
1. Typehint
1. Docstring

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+
+def get_bait_name(input_df):
+
```
```suggestion
```

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+
+def get_bait_name(input_df):
+
+    '''
+    input: DataFrame
+    output: string
+
+    Extracts the bait name from HS metrics DataFrame
+    '''
```
Please update docstring similar to read_hs_metrics example above.

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+    return qc_df
+
+
+def get_bait_name(input_df):
```
bait set name is known from config file. It can be extracted from config file. Have a look at one of the configs generated by BALSAMIC, and there is a field where filename exists.

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+    return str(bait_name[0])
+
+
+def get_qc_criteria(input_df, bait):
+
+    '''
+    input: DataFrame, string
+    output: DataFrame
+
+    Creates a new DataFrame with the QC criteria for only the deired bait set
+    '''
```
Similar to previous functions, please update:
1. Typehint
2. Docstring

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+    return qc_df
+
+
+def check_qc_criteria(input_qc_df, input_hsmetrics_df):
+
+    '''
+    input: DataFrame, DataFrame
+    output: DataFrame
+
+    This function can be devided in different parts:
+    1) Merging intersected values for the df with the desired QC criteria and bait set, with the HS Metrics df
+    2) Creating new columns with the QC-differences from the QC criteria
+    3) Setting QC flags
+    4) Extract the columns with the QC flag as a new DataFrame
+    '''
```
Same here. Similar to previous functions, please update:
1. Typehint
1. Docstring

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+    return qc_df
+
+
+def check_qc_criteria(input_qc_df, input_hsmetrics_df):
```
I'll add a general comment about this function in main review. In summary, it should just check values regardless of what it is. For example, one doesn't need to specifically ask for MEAN_TARGET_COVERAGE.

hassanfa commented on BALSAMIC/utils/qc_check.py:
```diff
@@ -0,0 +1,144 @@
+    print (qc_check_df)
+    return qc_check_df
+
+@click.command()
+@click.option('--hs_metrics', type = click.Path(exists=True), required = True, help = 'path to HS metrics for desired case')
+@click.option('--qc_table', type = click.Path(exists=True), required = True, help = 'path to qc table with criteria' )
+
+#The HS metrics and qc table provided in the command line will execute the main function.
+def main(hs_metrics, qc_table):
+
+
+    #Read the HS metrics and qc table and convert to df
+    hs_metrics_df = read_hs_metrics(hs_metrics)
+    qc_table_df = read_qc_table(qc_table)
+
+    #Extract the bait name and create a new df with the desired qc criteria
+    bait_set = get_bait_name(hs_metrics_df)
+    qc_criteria_df = get_qc_criteria(qc_table_df, bait_set)
+
+    #Create a df with qc-flag for each criteria for each sample
+    check_qc_criteria(qc_criteria_df, hs_metrics_df)
+
+
+if __name__ == '__main__':
+    main()
```
Great that you have this script with a Click functionality. It is preferable to have this as part of rest of BALSAMIC as well.

Perhaps it is better if you can remove Click, and have a `qc_check()`, so we can import it somewhere else?

Example:
```
from BALSAMIC.utils import qc_check

qc_check_summary  = qc_check(......)
```

This way we can use this script in a Snakemake rule and in CLI as well.

hassanfa reviewed (changes requested):
Nice work! I really liked the smart way of converting it to DataFrame instead. I added some general comments and some specific ones. We can use this script as it is, but it will be hard to integrate it to rest of BALSAMIC and Snakemake workflows.

Quality of code changes:

1. Update docstring and type hinting on the new functions
1. Add some unit tests so coverage doesn't drop more than 0.5%
1. Some code style change on empty lines etc (this can be done as last step of this pull request)


Logic and rational of what needs to be done:

1. Preferably this pull request should address the need for being able to check if an analysis has passed quality control metrics. These metrics can be diverse values for each panel, whole exome, or whole genome runs. Current implementation might make it challenging to add to rest of BALSAMIC. So it has to be a non-Click implementation.
1. The logic for `check_qc_criteria` is too specific. Expanding this feature should be easy in future. I suggest to just compare between the values set in `HSMETRICS_QC_CHECK` and actual input.

To address both points above I suggest the following:
Step 1. Find the path to `multiqc_picard_HsMetrics.json` in an analysis directory
Step 2. Convert it to a DataFrame
Step 3. Convert `HSMETRICS_QC_CHECK` to DataFrame
Step 4. Compare matching values in `HSMETRICS_QC_CHECK` DataFrame to values in `multiqc_picard_HsMetrics.json` DataFrame
Step 5. Extract Capture kit name from analysis config
Step 6. Filter out only matching column of capture kit with result from Step 4.

keyvanelhami commented:
I have now pushed following changes:

- Updated docstring and typehint for every function
- Rewrote the `check_qc_criteria` function according to comments from Hassan.
- Added two new functions: `failed_qc` and `write_output`

What remains:

- [ ] Rewrite `get_bait_name` to extract the bait name from config file instead
- [ ] Add unit test 
- [ ] Make the script compatible to run with Balsamic CLI by removing e.g. click 

@hassanfa please have a look and let me know if I have missed anything.

hassanfa commented:
@keyvanelhami Good plan!

keyvanelhami commented:
@hassanfa regarding the above bullet point `Rewrite get_bait_name to extract the bait name from config file instead`. Right now I am extracting this information from the hs_metrics.json. Why should we extract this information from the config instead? I can't see the point

keyvanelhami commented:
@hassanfa Please look into the final script. Also comment how I could improve my code if case is analyzed as single.

hassanfa commented:
Well done ✨✨

Let's fix:
1. autostyle using yapf/black in your pycharm, and good to go.
2. Some more test coverage. We are at 99%, would be cool to have it as close to it as possible.

It would be important to fix these for the next PR when you integrate this to the rest of BALSAMIC.

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,28 @@
+
+def test_read_hs_metrics():
+    #GIVEN the file exist
+    hs_metrics_path = "/Users/keyvan.elhami/Downloads/multiqc_picard_HsMetrics.json"
```
could you add these as fixture to `tests/test_data/qc_files/`? and then change this line?

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,28 @@
+
+def test_read_qc_table():
+    #GIVEN the file exist
+    qc_table_path = "/Users/keyvan.elhami/Downloads/qc_table4.json"
```
Same here: could you add these as fixture to `tests/test_data/qc_files/`? and then change this line?

hassanfa reviewed (changes requested):
Final changes and we are a go! Lets hope tests cover enough not to drop more than 1% 😁 Otherwise more tests 😁

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+        assert bol_list[n], "No values exists"
+
+
+def test_get_bait_and_sample_name():
```
use `tumor_normal_wgs_config ` fixture here.

```suggestion
def test_get_bait_and_sample_name(tumor_normal_wgs_config):
```

keyvanelhami replied:
@hassanfa should tumor_normal_wgs_config be a global variable? Or how should this value be provided in the function?

hassanfa replied:
`tumor_normal_wgs_config` is a `scope="session"` fixture. It is available throughout all tests: https://docs.pytest.org/en/stable/fixture.html#conftest-py-sharing-fixture-functions Fixtures are quite handy in testing stuff, mocking, etc.

There are four fixture for configs: `tumor_normal_config`, `tumor_only_config`, `tumor_normal_wgs_config`, and `tumor_only_wgs_config`. You can use them in every tests.

keyvanelhami replied:
Done!

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def test_get_bait_and_sample_name():
+    # GIVEN the file exists
+    config_file = "tests/test_data/qc_files/config_paired.json"
+
+    # WHEN reading the file
+    bed = get_bait_name(config_file)
+    sample_name = get_sample_name(config_file)
```
```suggestion
    # GIVEN a tumor_only case config
    # WHEN reading the file
    bed = get_bait_name(tumor_normal_wgs_config)
    sample_name = get_sample_name(tumor_normal_wgs_config)
```

keyvanelhami replied:
Done!

hassanfa commented on tests/test_data/qc_files/config_paired.json:
```diff
@@ -0,0 +1,151 @@
+{
```
could you rename this to case_config_tumor_normal.json ? It is more descriptive and matching.

keyvanelhami replied:
Done!

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+    assert len(nr_of_columns) == 2, "number of columns != 2"
+
+
+def test_check_qc_criteria_output_csv_and_qc():
+    # GIVEN following variables
+    config_file = "tests/test_data/qc_files/config_paired.json"
+    hs_metrics = "tests/test_data/qc_files/multiqc_picard_HsMetrics.json"
+    output_path = "tests/test_data/qc_files/output.csv"
```
Use PyTest's temporary path factory here:
```suggestion
def test_check_qc_criteria_output_csv_and_qc(tmp_path):
    # GIVEN following variables
    config_file = "tests/test_data/qc_files/config_paired.json"
    hs_metrics = "tests/test_data/qc_files/multiqc_picard_HsMetrics.json"
    
    test_new_dir = tmp_path / "check_qc_results" / "output.csv"
    test_new_dir.mkdir()
    output_path = test_new_dir / "config.json"
```

then `output_path.as_posix()` or `str(output_path)` is the one to use below in assert.

keyvanelhami replied:
Where is `tmp_path` provided when running the pytest?

hassanfa replied:
`tmp_path` is also a fixture, this is provided by pytest. very handy to create temporary path and files.

keyvanelhami replied:
Why do you use "config.json" in `output_path`? The `output_path` variable is used in the `write_output` function and should not be config-file, it should just indicate where to put the output.csv file

hassanfa replied:
My mistake, I meant "output.csv" or something similar.

So it should be:

```python
test_new_dir = tmp_path / "check_qc_results"
test_new_dir.mkdir()
output_path = Path(test_new_dir / "output.csv").as_posix()
```

hassanfa replied:
tmp_path and `mkdir()` build the temporary directory and return a `Path` object. `write_output` doesn't support `Path`objects, that's why I added `as_posix()`. One can use `str()` as well.

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+    # THEN check if the output df has 3 indexes, if csv-file exists and if qc check is string
+    nr_of_indexes = list(extract_criteria.index)
+    assert len(nr_of_indexes) == 3, "number of indexes != 3"
+    assert os.path.isfile(output_path), "File doesn't exists"
```
```suggestion
    assert output_path.exists(), "File doesn't exists"
```

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+
+    # THEN check if the output df has 3 indexes, if csv-file exists and if qc check is string
+    nr_of_indexes = list(extract_criteria.index)
+    assert len(nr_of_indexes) == 3, "number of indexes != 3"
```
Failing to test below will already break tests

```suggestion
    assert len(nr_of_indexes) == 3
```

keyvanelhami replied:
Done

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+from BALSAMIC.utils.qc_check import read_hs_metrics, read_qc_table, check_qc_criteria, write_output
+from BALSAMIC.utils.qc_check import get_bait_name, get_sample_name, get_qc_criteria, failed_qc
+from BALSAMIC.utils.constants import HSMETRICS_QC_CHECK
+import os
```
import pathlib
```suggestion
import os
from pathlib import Path
```

keyvanelhami replied:
Done

hassanfa commented on tests/utils/test_qc_check.py:
```diff
@@ -0,0 +1,82 @@
+
+def test_read_qc_table():
+    # GIVEN the file exist
+    try:
+        open("/BALSAMIC/utils/constants.py")
+    except IOError:
+        print("File doesn't exist")
```
Use import here instead of try/except
```python
from BALSAMIC.utils....
```

keyvanelhami replied:
Done

hassanfa reviewed (approved):
Wonderfully done Keyvan!

I particularly liked how the tests were arranged, `assert some_condisiton, "message"` was a nice touch, addition of two test json files was also smart.

Please fix the comments below and hit that merge button.

keyvanelhami commented:
@hassanfa Please have a final look and then merge :)

hassanfa commented:
Coverage 0.4%, but we need this feature so merging it.
